### PR TITLE
Added AniPlay provider and fixed M3u8Helper.selectBest

### PR DESCRIPTION
**Pull request by winstriker**
Added AniPlay.it italian provider, fixed M3u8Helper.selectBest function that was returning the worst quality instead of the best one

### CODE REVIEW

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        val json = app.get(url).text
+        val response = parseJson<ApiAnime>(json)
+
+        val tags: List<String> = response.genres.map { it.name }
+
+        val malId: Int? = response.websites.find { it.websiteId == 1 }?.url?.removePrefix("https://myanimelist.net/anime/")?.split("/")?.first()?.toIntOrNull()
+        val aniListId: Int? = response.websites.find { it.websiteId == 4 }?.url?.removePrefix("https://anilist.co/anime/")?.split("/")?.first()?.toIntOrNull()
+
+        val episodes : List<Episode>
+        if (response.seasons.isNullOrEmpty()) {
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt:161]: `val episodes = if ... else response.seasons.map{ it.toEpisodeList(url) }.flatten() `
flatten reduces a list of list to a single list
```
val a = listOf(1,2,3)
val b = listOf(4,5,6)
val c = listOf(a,b)
    
val d = c.flatten()
// d = 1,2,3,4,5,6

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    override suspend fun loadLinks(
+        data: String,
+        isCasting: Boolean,
+        subtitleCallback: (SubtitleFile) -> Unit,
+        callback: (ExtractorLink) -> Unit
+    ): Boolean {
+
+        val json = app.get(data).text
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt:194]: you can use parsed directly on `app.get(url).parsed<T>()`

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                this.quality = if (it.fullHD) SearchQuality.HD else null
+            }
+        }
+        return HomePageResponse(listOf(HomePageList("Ultime uscite",results)))
+    }
+
+    override suspend fun search(query: String): List<SearchResponse> {
+        val json = app.get("$mainUrl/api/anime/advanced-search?page=0&size=36&query=$query").text
+        val response = parseJson<List<ApiSearchResult>>(json)
+        val results = response.map {
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt:135]: you can return directly return response.map

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    data class ApiEpisodeUrl(
+        @JsonProperty("videoUrl") val url: String
+    )
+
+    override suspend fun getMainPage(): HomePageResponse {
+        val json = app.get("$mainUrl/api/home/latest-episodes?page=0").text
+        val response = parseJson<List<ApiMainPageAnime>>(json)
+        val results = response.map{
+            newAnimeSearchResponse(
+                name = if (it.title.contains(" (ITA)")) it.title.replace(" (ITA)", "") else it.title,
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AniPlayProvider.kt:120]: you are checking for (ITA) a lot, might be good to abstract away to const variable or function?

**Review by LagradOst — APPROVED** [review]
Very good work 🥳

minor code changes for better code quality

**Comment by winstriker** [comment]
Done! Thanks again for your changes, I didn't know of flatten and I think it'll be useful in other occasions too.

**Pull request merged**